app/views/directives/key-value-editor.html

```diff
@@ -0,0 +1,239 @@
+          <span ng-if="entry.keyValidatorErrorTooltip || keyValidatorErrorTooltip" class="help action-inline">
+            <a
+              aria-hidden="true"
+              data-toggle="tooltip"
+              data-placement="top"
+              data-original-title="{{entry.keyValidatorErrorTooltip || keyValidatorErrorTooltip}}"
+              title="{{entry.keyValidatorErrorTooltip || keyValidatorErrorTooltip}}">
+              <i class="{{entry.keyValidatorErrorTooltipIcon || keyValidatorErrorTooltipIcon}}"></i>
+            </a>
+          </span>
+        </span>
+        <span
+          class="help-block key-min-length"
+          ng-show="(forms.keyValueEditor[uniqueForKey(unique, $index)].$error.minlength)">
+          <span class="validation-text">Minimum character count is {{keyMinlength}}</span>
+        </span>
+        <span
+          class="help-block key-validation-error"
+          ng-show="(forms.keyValueEditor[uniqueForKey(unique, $index)].$error.required)">
+          <span class="validation-text">{{keyRequiredError}}</span>
+        </span>
+      </div>
+
+      <!-- the value block -->
+      <div
+        class="form-group key-value-editor-input"
+        ng-class="forms.keyValueEditor[uniqueForValue(unique, $index)].$invalid ? 'has-error' : ''">
+
+        <label for="uniqueForValue(unique, $index)" class="sr-only">{{valuePlaceholder}}</label>
+
+        <div ng-if="(!entry.valueFrom)">
+          <input
+            ng-if="(!entry.valueAlt)"
+            type="text"
+            class="form-control"
+            ng-class="{ '{{setFocusValClass}}' : $last  }"
+            id="{{uniqueForValue(unique, $index)}}"
+            name="{{uniqueForValue(unique, $index)}}"
+            ng-attr-placeholder="{{ (!isReadonlyAny) && valuePlaceholder || ''}}"
+            ng-minlength="{{valueMinlength}}"
+            maxlength="{{valueMaxlength}}"
+            ng-model="entry.value"
+            ng-readonly="isReadonlyAny || isReadonlySome(entry.name) || entry.isReadonly"
+            ng-pattern="validation.val"
+            ng-required="!allowEmptyKeys && entry.value">
+        </div>
+
+        <div ng-if="entry.valueFrom">
+          <div ng-if="isReadonlyAny || entry.isReadonlyValue" class="faux-input-group">
+            <span
+              class="faux-form-control-addon {{entry.valueIcon}}"
+              aria-hidden="true"
+              data-toggle="tooltip"
+              data-placement="top"
+              data-original-title="{{entry.valueIconTooltip || valueIconTooltip}}"
+              title="{{entry.valueIconTooltip || valueIconTooltip}}"></span>
+            <div
+              class="faux-form-control readonly">
+              <span ng-switch="entry.refType">
+                <span ng-switch-when="configMapKeyRef">
+                  Set to the key {{entry.valueFrom.configMapKeyRef.key}} in config map
+                  <span
+                    ng-if="entry.isReadonlyValue">
```

Comment:
We are [using canI to check if the user has rights to view config maps or secrets](https://github.com/rhamilto/origin-web-console/blob/3940968a6e3e5ab3b881f396788b85236c5f746c/app/scripts/directives/keyValueEditor.js#L197-L198), we're just using a single scope parameter (entry.isReadonlyValue) to track multiple conditions as this was easier than having multiple params.  The thought being it was better to keep this logic in the controller and have one less key to clean off entry.  Yea/nay?

1.  If the user can get config maps or secrets and the kve instance is NOT read only, we show the ui-selects for config maps and secrets.

![screen shot 2017-02-15 at 9 03 35 am](https://cloud.githubusercontent.com/assets/895728/22977689/b72ac5d8-f35d-11e6-8587-c951c7cd9917.PNG)

2.  If the user can get config maps or secrets and the kve instance is read only, we show the faux input with the link.

![screen shot 2017-02-15 at 9 03 46 am](https://cloud.githubusercontent.com/assets/895728/22977703/c39d5182-f35d-11e6-8fdc-c522c7b97fc7.PNG)

3.  If the user can NOT get config maps or secrets and the kve instance is NOT read only, we should the faux input without the link.

![screen shot 2017-02-15 at 9 11 42 am](https://cloud.githubusercontent.com/assets/895728/22977982/cf8ee996-f35e-11e6-81be-dccb91047541.PNG)

4.  If the user can NOT get config maps or secrets and the kve instance is read only, we should the faux input without the link.

![screen shot 2017-02-15 at 9 11 36 am](https://cloud.githubusercontent.com/assets/895728/22977990/d6fd22f6-f35e-11e6-9c78-151da84646df.PNG)